File: api/library/comprehensive/engine.py
# =============================================================================
# COMPREHENSIVE BAZI ENGINE — Main Entry Point
# =============================================================================
# Provides the primary interface for constructing ChartData from pillar inputs
# and generating complete reports.
# =============================================================================

from typing import List, Dict, Optional, Tuple
from ..core import STEMS, BRANCHES
from ..derived import get_all_branch_qi, STEM_ORDER, BRANCH_ORDER
from .models import ChartData, Pillar, LuckPillarInfo
from .report import generate_comprehensive_report


def build_chart(
    gender: str,
    birth_year: int,
    year_stem: str, year_branch: str,
    month_stem: str, month_branch: str,
    day_stem: str, day_branch: str,
    hour_stem: str, hour_branch: str,
    luck_pillar_stem: str = "",
    luck_pillar_branch: str = "",
    luck_pillars: Optional[List[dict]] = None,
    current_year: int = 0,
    annual_stem: str = "", annual_branch: str = "",
    monthly_stem: str = "", monthly_branch: str = "",
    daily_stem: str = "", daily_branch: str = "",
    hourly_stem: str = "", hourly_branch: str = "",
) -> ChartData:
    """
    Build a ChartData object from raw pillar inputs.

    Args:
        gender: "male" or "female"
        birth_year: Gregorian birth year
        year/month/day/hour_stem: Pinyin stem IDs (e.g., "Jia", "Yi")
        year/month/day/hour_branch: Pinyin branch IDs (e.g., "Zi", "Chou")
        luck_pillar_stem/branch: Current luck pillar (optional)
        luck_pillars: List of dicts with keys: stem, branch, start_age, end_age, start_year, end_year
        current_year: Override current year (0 = auto-detect)
    """
    from datetime import datetime
    if current_year == 0:
        current_year = datetime.now().year

    age = current_year - birth_year

    # Build pillar objects
    pillars = {}
    for pos, stem, branch in [
        ("year", year_stem, year_branch),
        ("month", month_stem, month_branch),
        ("day", day_stem, day_branch),
        ("hour", hour_stem, hour_branch),
    ]:
        qi = get_all_branch_qi(branch)
        pillars[pos] = Pillar(
            position=pos,
            stem=stem,
            branch=branch,
            stem_chinese=STEMS[stem]["chinese"],
            branch_chinese=BRANCHES[branch]["chinese"],
            stem_element=STEMS[stem]["element"],
            stem_polarity=STEMS[stem]["polarity"],
            branch_element=BRANCHES[branch]["element"],
            branch_polarity=BRANCHES[branch]["polarity"],
            hidden_stems=qi,
        )

    # Build luck pillar
    lp = None
    if luck_pillar_stem and luck_pillar_branch:
        lp_qi = get_all_branch_qi(luck_pillar_branch)
        lp = Pillar(
            position="luck_pillar",
            stem=luck_pillar_stem,
            branch=luck_pillar_branch,
            stem_chinese=STEMS[luck_pillar_stem]["chinese"],
            branch_chinese=BRANCHES[luck_pillar_branch]["chinese"],
            stem_element=STEMS[luck_pillar_stem]["element"],
            stem_polarity=STEMS[luck_pillar_stem]["polarity"],
            branch_element=BRANCHES[luck_pillar_branch]["element"],
            branch_polarity=BRANCHES[luck_pillar_branch]["polarity"],
            hidden_stems=lp_qi,
        )

    # Build luck pillars list
    lp_list = []
    if luck_pillars:
        for lp_data in luck_pillars:
            from ..derived import get_ten_god
            stem = lp_data["stem"]
            branch = lp_data["branch"]
            tg = get_ten_god(day_stem, stem)
            lp_info = LuckPillarInfo(
                stem=stem,
                branch=branch,
                stem_chinese=STEMS[stem]["chinese"],
                branch_chinese=BRANCHES[branch]["chinese"],
                start_age=lp_data.get("start_age", 0),
                end_age=lp_data.get("end_age", 0),
                start_year=lp_data.get("start_year", 0),
                end_year=lp_data.get("end_year", 0),
                stem_ten_god=tg[0] if tg else "",
                stem_ten_god_chinese=tg[2] if tg else "",
                is_current=(luck_pillar_stem == stem and luck_pillar_branch == branch),
                hidden_stems=get_all_branch_qi(branch),
            )
            lp_list.append(lp_info)

    # Build time-period pillars
    tp_pillars = {}
    for tp_pos, tp_stem, tp_branch in [
        ("annual", annual_stem, annual_branch),
        ("monthly", monthly_stem, monthly_branch),
        ("daily", daily_stem, daily_branch),
        ("hourly", hourly_stem, hourly_branch),
    ]:
        if tp_stem and tp_branch:
            tp_qi = get_all_branch_qi(tp_branch)
            tp_pillars[tp_pos] = Pillar(
                position=tp_pos,
                stem=tp_stem,
                branch=tp_branch,
                stem_chinese=STEMS[tp_stem]["chinese"],
                branch_chinese=BRANCHES[tp_branch]["chinese"],
                stem_element=STEMS[tp_stem]["element"],
                stem_polarity=STEMS[tp_stem]["polarity"],
                branch_element=BRANCHES[tp_branch]["element"],
                branch_polarity=BRANCHES[tp_branch]["polarity"],
                hidden_stems=tp_qi,
            )

    # Day Master info
    dm = day_stem
    dm_info = STEMS[dm]

    # Current year stems
    cy_stem_idx = (current_year - 4) % 10
    cy_branch_idx = (current_year - 4) % 12

    chart = ChartData(
        gender=gender.lower(),
        birth_year=birth_year,
        age=age,
        pillars=pillars,
        day_master=dm,
        dm_element=dm_info["element"],
        dm_polarity=dm_info["polarity"],
        dm_chinese=dm_info["chinese"],
        luck_pillar=lp,
        luck_pillars=lp_list,
        time_period_pillars=tp_pillars,
        current_year_stem=STEM_ORDER[cy_stem_idx],
        current_year_branch=BRANCH_ORDER[cy_branch_idx],
    )

    return chart


def analyze_for_api(chart: ChartData) -> dict:
    """
    Run full analysis and return all intermediate results (not just markdown).
    Used by the /api/analyze_bazi adapter.
    """
    from .ten_gods import map_all_ten_gods, classify_ten_god_strength
    from .strength import assess_day_master_strength
    from .interactions import detect_all_interactions
    from .shen_sha import run_all_shen_sha
    from .predictions import run_all_predictions
    from .environment import assess_environment

    strength = assess_day_master_strength(chart)
    tg_entries = map_all_ten_gods(chart)
    tg_classification = classify_ten_god_strength(tg_entries)
    interactions = detect_all_interactions(chart)
    shen_sha = run_all_shen_sha(chart)
    predictions = run_all_predictions(chart)
    env = assess_environment(chart, strength)
    comprehensive_report = generate_comprehensive_report(chart)

    return {
        "strength": strength,
        "ten_god_entries": tg_entries,
        "ten_god_classification": tg_classification,
        "interactions": interactions,
        "shen_sha": shen_sha,
        "predictions": predictions,
        "environment": env,
        "comprehensive_report": comprehensive_report,
    }


def analyze(
    gender: str,
    birth_year: int,
    year_stem: str, year_branch: str,
    month_stem: str, month_branch: str,
    day_stem: str, day_branch: str,
    hour_stem: str, hour_branch: str,
    luck_pillar_stem: str = "",
    luck_pillar_branch: str = "",
    luck_pillars: Optional[List[dict]] = None,
    current_year: int = 0,
) -> str:
    """
    One-call comprehensive BaZi analysis.
    Returns complete markdown report.
    """
    chart = build_chart(
        gender=gender,
        birth_year=birth_year,
        year_stem=year_stem, year_branch=year_branch,
        month_stem=month_stem, month_branch=month_branch,
        day_stem=day_stem, day_branch=day_branch,
        hour_stem=hour_stem, hour_branch=hour_branch,
        luck_pillar_stem=luck_pillar_stem,
        luck_pillar_branch=luck_pillar_branch,
        luck_pillars=luck_pillars,
        current_year=current_year,
    )
    return generate_comprehensive_report(chart)

File: src/lib/bazi/comprehensive/engine.ts
import 'server-only';

// =============================================================================
// COMPREHENSIVE BAZI ENGINE — Main Entry Point
// =============================================================================
// Provides the primary interface for constructing ChartData from pillar inputs
// and generating complete reports.
// =============================================================================

import { STEMS, BRANCHES, type StemName, type BranchName } from '../core';
import { getAllBranchQi, getTenGod, STEM_ORDER, BRANCH_ORDER } from '../derived';
import type { ChartData, Pillar, LuckPillarInfo, StrengthAssessment } from './models';
import { generateComprehensiveReport } from './report';
import { mapAllTenGods, classifyTenGodStrength } from './ten-gods';
import { assessDayMasterStrength } from './strength';
import { detectAllInteractions } from './interactions';
import { runAllShenSha } from './shen-sha';
import { runAllPredictions } from './predictions';
import { assessEnvironment } from './environment';
import { calculateWuxing, type WuxingInput, type WuxingResult } from '../wuxing/calculator';


export function buildChart(args: {
  gender: string;
  birth_year: number;
  year_stem: string; year_branch: string;
  month_stem: string; month_branch: string;
  day_stem: string; day_branch: string;
  hour_stem: string; hour_branch: string;
  luck_pillar_stem?: string;
  luck_pillar_branch?: string;
  luck_pillars?: Array<Record<string, unknown>>;
  current_year?: number;
  annual_stem?: string; annual_branch?: string;
  monthly_stem?: string; monthly_branch?: string;
  daily_stem?: string; daily_branch?: string;
  hourly_stem?: string; hourly_branch?: string;
}): ChartData {
  /**
   * Build a ChartData object from raw pillar inputs.
   */
  const {
    gender, birth_year,
    year_stem, year_branch,
    month_stem, month_branch,
    day_stem, day_branch,
    hour_stem, hour_branch,
    luck_pillar_stem = "",
    luck_pillar_branch = "",
    luck_pillars,
    annual_stem = "", annual_branch = "",
    monthly_stem = "", monthly_branch = "",
    daily_stem = "", daily_branch = "",
    hourly_stem = "", hourly_branch = "",
  } = args;

  let currentYear = args.current_year ?? 0;
  if (currentYear === 0) {
    currentYear = new Date().getFullYear();
  }

  const age = currentYear - birth_year;

  // Build pillar objects
  const pillars: Record<string, Pillar> = {};
  const pillarDefs: Array<[string, string, string]> = [
    ["year", year_stem, year_branch],
    ["month", month_stem, month_branch],
    ["day", day_stem, day_branch],
    ["hour", hour_stem, hour_branch],
  ];

  for (const [pos, stem, branch] of pillarDefs) {
    const qi = getAllBranchQi(branch);
    pillars[pos] = {
      position: pos,
      stem: stem as StemName,
      branch: branch as BranchName,
      stem_chinese: STEMS[stem as StemName].chinese,
      branch_chinese: BRANCHES[branch as BranchName].chinese,
      stem_element: STEMS[stem as StemName].element,
      stem_polarity: STEMS[stem as StemName].polarity,
      branch_element: BRANCHES[branch as BranchName].element,
      branch_polarity: BRANCHES[branch as BranchName].polarity,
      hidden_stems: [...qi] as Array<[StemName, number]>,
    };
  }

  // Build luck pillar
  let lp: Pillar | null = null;
  if (luck_pillar_stem && luck_pillar_branch) {
    const lpQi = getAllBranchQi(luck_pillar_branch);
    lp = {
      position: "luck_pillar",
      stem: luck_pillar_stem as StemName,
      branch: luck_pillar_branch as BranchName,
      stem_chinese: STEMS[luck_pillar_stem as StemName].chinese,
      branch_chinese: BRANCHES[luck_pillar_branch as BranchName].chinese,
      stem_element: STEMS[luck_pillar_stem as StemName].element,
      stem_polarity: STEMS[luck_pillar_stem as StemName].polarity,
      branch_element: BRANCHES[luck_pillar_branch as BranchName].element,
      branch_polarity: BRANCHES[luck_pillar_branch as BranchName].polarity,
      hidden_stems: [...lpQi] as Array<[StemName, number]>,
    };
  }

  // Build luck pillars list
  const lpList: LuckPillarInfo[] = [];
  if (luck_pillars) {
    for (const lpData of luck_pillars) {
      const stem = lpData.stem as string;
      const branch = lpData.branch as string;
      const tg = getTenGod(day_stem as StemName, stem as StemName);
      lpList.push({
        stem: stem as StemName,
        branch: branch as BranchName,
        stem_chinese: STEMS[stem as StemName].chinese,
        branch_chinese: BRANCHES[branch as BranchName].chinese,
        start_age: (lpData.start_age as number) ?? 0,
        end_age: (lpData.end_age as number) ?? 0,
        start_year: (lpData.start_year as number) ?? 0,
        end_year: (lpData.end_year as number) ?? 0,
        stem_ten_god: tg ? tg[0] : "",
        stem_ten_god_chinese: tg ? tg[2] : "",
        is_current: luck_pillar_stem === stem && luck_pillar_branch === branch,
        hidden_stems: [...getAllBranchQi(branch)] as Array<[StemName, number]>,
      });
    }
  }

  // Build time-period pillars
  const tpPillars: Record<string, Pillar> = {};
  const tpDefs: Array<[string, string, string]> = [
    ["annual", annual_stem, annual_branch],
    ["monthly", monthly_stem, monthly_branch],
    ["daily", daily_stem, daily_branch],
    ["hourly", hourly_stem, hourly_branch],
  ];
  for (const [tpPos, tpStem, tpBranch] of tpDefs) {
    if (tpStem && tpBranch) {
      const tpQi = getAllBranchQi(tpBranch);
      tpPillars[tpPos] = {
        position: tpPos,
        stem: tpStem as StemName,
        branch: tpBranch as BranchName,
        stem_chinese: STEMS[tpStem as StemName].chinese,
        branch_chinese: BRANCHES[tpBranch as BranchName].chinese,
        stem_element: STEMS[tpStem as StemName].element,
        stem_polarity: STEMS[tpStem as StemName].polarity,
        branch_element: BRANCHES[tpBranch as BranchName].element,
        branch_polarity: BRANCHES[tpBranch as BranchName].polarity,
        hidden_stems: [...tpQi] as Array<[StemName, number]>,
      };
    }
  }

  // Day Master info
  const dm = day_stem as StemName;
  const dmInfo = STEMS[dm];

  // Current year stems
  const cyStemIdx = ((currentYear - 4) % 10 + 10) % 10;
  const cyBranchIdx = ((currentYear - 4) % 12 + 12) % 12;

  const chart: ChartData = {
    gender: gender.toLowerCase(),
    birth_year,
    age,
    pillars,
    day_master: dm,
    dm_element: dmInfo.element,
    dm_polarity: dmInfo.polarity,
    dm_chinese: dmInfo.chinese,
    luck_pillar: lp,
    luck_pillars: lpList,
    time_period_pillars: tpPillars,
    current_year_stem: STEM_ORDER[cyStemIdx],
    current_year_branch: BRANCH_ORDER[cyBranchIdx],
  };

  return chart;
}


export function analyzeForApi(chart: ChartData): Record<string, unknown> {
  /**
   * Run full analysis and return all intermediate results (not just markdown).
   * Used by the /api/analyze_bazi adapter.
   */
  const interactions = detectAllInteractions(chart);

  // Old strength assessment (kept for backward-compatible fields)
  const oldStrength = assessDayMasterStrength(chart, interactions);

  // New Wu Xing calculator
  const wuxingInput: WuxingInput = {
    yearPillar: { stem: chart.pillars.year.stem, branch: chart.pillars.year.branch },
    monthPillar: { stem: chart.pillars.month.stem, branch: chart.pillars.month.branch },
    dayPillar: { stem: chart.pillars.day.stem, branch: chart.pillars.day.branch },
    hourPillar: { stem: chart.pillars.hour.stem, branch: chart.pillars.hour.branch },
    age: chart.age,
    gender: chart.gender === 'male' ? 'M' : 'F',
    location: 'hometown',
  };
  const wuxingResult = calculateWuxing(wuxingInput);

  // Map wuxing strength label to legacy verdict
  const STRENGTH_TO_VERDICT: Record<string, string> = {
    dominant: 'extremely_strong',
    strong: 'strong',
    balanced: 'neutral',
    weak: 'weak',
    very_weak: 'extremely_weak',
  };

  // Build element_percentages from wuxing result
  const wuxingPercentages: Record<string, number> = {};
  for (const elem of ['Wood', 'Fire', 'Earth', 'Metal', 'Water'] as const) {
    wuxingPercentages[elem] = wuxingResult.elements[elem].percent;
  }

  // Build merged StrengthAssessment: wuxing values override, old fields kept for compat
  const strength: StrengthAssessment = {
    score: wuxingResult.dayMaster.percent,
    verdict: STRENGTH_TO_VERDICT[wuxingResult.dayMaster.strength] ?? 'neutral',
    useful_god: wuxingResult.gods.useful,
    element_percentages: wuxingPercentages,
    favorable_elements: [wuxingResult.gods.useful, wuxingResult.gods.favorable],
    unfavorable_elements: [wuxingResult.gods.unfavorable, wuxingResult.gods.enemy],
    // Backward-compatible fields from old assessment
    support_count: oldStrength.support_count,
    drain_count: oldStrength.drain_count,
    seasonal_state: oldStrength.seasonal_state,
    is_following_chart: oldStrength.is_following_chart,
    following_type: oldStrength.following_type,
    best_element_pairs: oldStrength.best_element_pairs,
  };

  const tgEntries = mapAllTenGods(chart);
  const tgClassification = classifyTenGodStrength(tgEntries);
  const shenSha = runAllShenSha(chart);
  const predictions = runAllPredictions(chart);
  const env = assessEnvironment(chart, strength);
  const comprehensiveReport = generateComprehensiveReport(chart);

  return {
    strength,
    wuxing_result: wuxingResult,
    ten_god_entries: tgEntries,
    ten_god_classification: tgClassification,
    interactions,
    shen_sha: shenSha,
    predictions,
    environment: env,
    comprehensive_report: comprehensiveReport,
  };
}


export function analyze(args: {
  gender: string;
  birth_year: number;
  year_stem: string; year_branch: string;
  month_stem: string; month_branch: string;
  day_stem: string; day_branch: string;
  hour_stem: string; hour_branch: string;
  luck_pillar_stem?: string;
  luck_pillar_branch?: string;
  luck_pillars?: Array<Record<string, unknown>>;
  current_year?: number;
}): string {
  /**
   * One-call comprehensive BaZi analysis.
   * Returns complete markdown report.
   */
  const chart = buildChart(args);
  return generateComprehensiveReport(chart);
}
